pkg/binary/envoy/runtime.go

```diff
@@ -104,17 +104,28 @@ type Runtime struct {
 	preStart       []func(binary.Runner) error
 	preTermination []func(binary.Runner) error
 
-	isReady bool
+	getAdminAddress func(r *Runtime) string
+	isReady         bool
+}
+
+// GetAdminAddress returns the current admin address in host:port format, or empty if not yet available.
+func (r *Runtime) GetAdminAddress() string {
+	if r.getAdminAddress == nil {
+		return r.Config.GetAdminAddress()
+	}
+	return r.getAdminAddress(r)
 }
 
 // Status indicates the state of the child process
 func (r *Runtime) Status() int {
 	switch {
 	case r.cmd == nil, r.cmd.Process == nil:
 		return binary.StatusStarting
-	case r.cmd.ProcessState == nil:
-		if r.envoyReady() {
-			return binary.StatusReady
+	case r.cmd.ProcessState == nil: // the process started, but it hasn't completed, yet
+		// TODO: envoyReady() can succeed when there's a port collision even when the process managed by this
```

Comment:
this took me a lot of head scratching to eventually understand, hence the comment now.